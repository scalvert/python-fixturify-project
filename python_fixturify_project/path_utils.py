from pathlib import Path


def write_to_file(file_path: str, contents: str) -> None:
    with open(Path(file_path), "w") as f:
        f.write(contents)


def create_directory(file_path: str) -> None:
    Path(file_path).mkdir(parents=True, exist_ok=True)
