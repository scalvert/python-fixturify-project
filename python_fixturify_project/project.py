from typing import Any, Dict

import glob
import os
import shutil
import tempfile

from dict_path import extract_dict, inject_dict
from wcmatch.pathlib import DOTGLOB, GLOBSTAR, Path

from python_fixturify_project.exceptions import InvalidProjectError
from python_fixturify_project.path_utils import create_directory, write_to_file
from python_fixturify_project.utils import deep_merge, keys_exists


class Project:
    def __init__(self, base_dir=None, files=None):
        self._base_dir = base_dir
        self._files = files or {}

        self.write(self._files)

    def __del__(self):
        # Ensure we clean up the temp dir structure on delete
        self.__exit__()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        try:
            shutil.rmtree(self.base_dir)
        except FileNotFoundError:
            # No need to do anything, file structure has already been cleaned up!
            pass

    @property
    def base_dir(self):
        """Gets the base directory path, usually a tmp directory unless a baseDir has been explicitly set."""
        return self._base_dir

    @base_dir.setter
    def base_dir(self, value):
        """Sets the base directory of the project."""
        self._base_dir = value

    @base_dir.deleter
    def base_dir(self):
        """Deletes the base directory."""
        del self._base_dir

    @property
    def files(self):
        """Gets the files containing the project's directory structure."""
        return self._files

    @files.setter
    def files(self, value):
        """Sets the files corresponding to the project's directory structure."""
        self._files = value

    @files.deleter
    def files(self):
        """Deletes the files corresponding to the project's directory structure."""
        del self._files

    def __auto_base_dir(self):
        """Creates and sets the base_dir if not explicitly configured during init"""
        if not self.base_dir:
            self.base_dir = tempfile.mkdtemp()

    def merge_files(self, dir_json):
        """Merges an object containing a directory represention with the existing files."""
        self.files = deep_merge(self.files, dir_json)

    def write(self, dir_json):
        """Writes the existing files property containing a directory representation to the tmp directory."""
        if dir_json:
            self.merge_files(dir_json)
        self.__write_project()

    def read(self, ignore_patterns=["**/.git", "**/.git/**"]):
        """Reads the contents of the base_dir to a dict and ignores any files/dirs matched by the glob expressions"""
        files: Dict[str, Any] = {}

        for path in Path(self.base_dir).rglob(
            "*", exclude=ignore_patterns, flags=DOTGLOB | GLOBSTAR
        ):
            rel_path = path.relative_to(self.base_dir)

            if str(rel_path) == ".":
                continue

            if path.is_file():
                with open(path) as f:
                    self.__add_file(files, rel_path, f.read())
            else:
                self.__add_dir(files, rel_path)

        return files

    def get(self, object_path: str):
        return extract_dict(self._files, object_path)

    def __add_file(self, files, path, contents):
        file = os.path.basename(path)
        dir_name = os.path.dirname(path)

        if dir_name != "":
            self.__add_dir(files, dir_name)[file] = contents
        else:
            files[file] = contents

    def __add_dir(self, files, path):
        path = str(path)
        if not keys_exists(files, *path.split("/")):
            inject_dict(files, path, {})

        return extract_dict(files, path)

    def __write_project(self):
        self.__auto_base_dir()
        self.__write(self.files, self.base_dir)

    def __write(self, dir_structure, full_path):
        """Recursive write helper function. Does the bulk of the work in terms of writing the directory structure"""
        # Base case
        if not dir_structure or not isinstance(dir_structure, dict):
            return

        # Save the original path
        original_dir = Path(full_path)
        for entry in dir_structure:
            full_path = Path(full_path, entry)
            if not isinstance(entry, str) or entry == "":
                raise InvalidProjectError(
                    "Invalid directory structure given. Each key must be a non-empty string"
                )
            if isinstance(dir_structure[entry], str):  # This is a file
                write_to_file(full_path, dir_structure[entry])
            else:
                if entry == "." or entry == "..":
                    raise InvalidProjectError('Directory entry must not be "." or ".."')

                create_directory(full_path)
                # Our recursion step, which should only happen if we find ourselves a nested directory
                self.__write(dir_structure=dir_structure[entry], full_path=full_path)

            # Reset the original path because Python will still remember the value of `full_path` even after we return from recursion
            full_path = original_dir
