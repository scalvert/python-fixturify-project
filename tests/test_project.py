from distutils.dir_util import copy_tree
from os import path
from pathlib import Path

import pytest
from conftest import BAD_DIR_NAME, BAD_EMPTY_NAME, GOOD_NESTED_DIRS, GOOD_SINGLE_FILE
from deepdiff import DeepDiff

from python_fixturify_project.exceptions import InvalidProjectError
from python_fixturify_project.project import Project


def test_project():
    """Example test with parametrization."""
    project = Project()
    assert isinstance(project, Project)


def test_project_has_base_dir_on_instantiation():
    project = Project()
    assert type(project.base_dir) is str


def test_cleanup_dir():
    project = Project()
    project.write(GOOD_SINGLE_FILE)
    base_dir = project.base_dir

    assert path.exists(base_dir) and path.isdir(base_dir)

    del project

    assert not path.exists(base_dir)


def test_get_files_after_copy(snapshot):
    project1 = Project(
        files={
            "one.py": "# some python",
            "dir": {
                "two.py": "# another python",
                "dir2": {"three.py": "# and this makes 3!!!"},
            },
        }
    )

    project2 = Project()

    copy_tree(project1.base_dir, project2.base_dir)

    assert project2.files == snapshot


def test_get_from_files(snapshot):
    project = Project(
        files={
            "one.py": "# some python",
            "dir": {
                "two.py": "# another python",
                "dir2": {"three.py": "# and this makes 3!!!"},
            },
        }
    )

    assert project.get("one.py") == snapshot(name="one.py")
    assert project.get("dir") == snapshot(name="dir")
    assert project.get("dir/dir2") == snapshot(name="dir2")


def test_get_from_copied_files(snapshot):
    project1 = Project(
        files={
            "one.py": "# some python",
            "dir": {
                "two.py": "# another python",
                "dir2": {"three.py": "# and this makes 3!!!"},
            },
        }
    )

    project2 = Project()

    copy_tree(project1.base_dir, project2.base_dir)

    project2.get("one.py")

    assert project2.files == snapshot
    assert DeepDiff(project1.files, project2.files) == {}


@pytest.mark.parametrize("test_input", [BAD_DIR_NAME, BAD_EMPTY_NAME])
def test_improper_write(test_input):
    with pytest.raises(InvalidProjectError):
        project = Project()
        project.write(test_input)


def test_multiple_writes_correctly_merges(snapshot):
    project = Project(files=GOOD_NESTED_DIRS)

    assert project.files == snapshot(name="original_files")

    project.write({
        'another.py': 'Yet another!!!'
    })

    assert project.files == snapshot(name="merged_files")


def test_proper_write_with_dispose(snapshot):
    base_dir = None

    project = Project(files=GOOD_NESTED_DIRS)

    base_dir = project.base_dir

    assert path.exists(base_dir) and path.isdir(base_dir)

    assert project.read() == snapshot

    project.dispose()

    assert not path.exists(base_dir)


def test_read_recreates_project_from_disc(snapshot):
    project = Project(files=GOOD_NESTED_DIRS)

    dir_json = project.read()

    assert dir_json == snapshot


def test_read_recreates_project_from_disc_with_similar_filenames(snapshot):
    files = {
        "valid_file.txt": "some text",
        "sub": {
            "valid_file.txt": "some text",
            "sub": {
                "valid_file.txt": "some text",
            },
        },
    }

    project = Project(files=files)

    dir_json = project.read()

    assert dir_json == snapshot


def test_read_ignore_files(snapshot):
    files = {
        ".git": {"a_nested_dir": {"last_nested_dir": {"a_file": "some text"}}},
        ".github": {"ignore_me": {}, "do_not_ignore_me": {"a_file": "some text"}},
        "ignore_me": "some text",
        "do_not_ignore_me": "some text",
    }

    project = Project(files=files, ignore_patterns=["**/ignore_me"])

    dir_json = project.read()

    assert dir_json == snapshot
