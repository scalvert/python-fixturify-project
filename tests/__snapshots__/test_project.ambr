# name: test_get_from_files[dir2]
  dict({
    'three.py': '# and this makes 3!!!',
  })
# ---
# name: test_get_from_files[dir]
  dict({
    'dir2': dict({
      'three.py': '# and this makes 3!!!',
    }),
    'two.py': '# another python',
  })
# ---
# name: test_get_from_files[one.py]
  '# some python'
# ---
# name: test_proper_write_with_cleanup
  dict({
    '.a_hidden_folder': dict({
      'nested_dir': dict({
        '.a_hidden_file': 'some text',
      }),
    }),
    'nested_dir': dict({
      'another_nested_dir': dict({
        'final_text_file.txt': 'some text',
        'last_nested_empty_dir': dict({
        }),
      }),
      'another_nested_empty_dir': dict({
      }),
      'valid_empty_file.txt': '',
    }),
    'valid_file.txt': 'some text',
  })
# ---
# name: test_read_ignore_files
  dict({
    '.github': dict({
      'do_not_ignore_me': dict({
        'a_file': 'some text',
      }),
    }),
    'do_not_ignore_me': 'some text',
  })
# ---
# name: test_read_recreates_project_from_disc
  dict({
    '.a_hidden_folder': dict({
      'nested_dir': dict({
        '.a_hidden_file': 'some text',
      }),
    }),
    'nested_dir': dict({
      'another_nested_dir': dict({
        'final_text_file.txt': 'some text',
        'last_nested_empty_dir': dict({
        }),
      }),
      'another_nested_empty_dir': dict({
      }),
      'valid_empty_file.txt': '',
    }),
    'valid_file.txt': 'some text',
  })
# ---
# name: test_read_recreates_project_from_disc_with_similar_filenames
  dict({
    'sub': dict({
      'sub': dict({
        'valid_file.txt': 'some text',
      }),
      'valid_file.txt': 'some text',
    }),
    'valid_file.txt': 'some text',
  })
# ---
